File: backend/utils/file_processing.py
"""Extract text from uploaded documents (PDF, DOCX, TXT)."""
import os
from pathlib import Path


def extract_text_from_pdf(file_path: str) -> str:
    import fitz  # PyMuPDF
    doc = fitz.open(file_path)
    text_parts = []
    for page in doc:
        text_parts.append(page.get_text())
    doc.close()
    return "\n".join(text_parts)


def extract_text_from_docx(file_path: str) -> str:
    from docx import Document
    doc = Document(file_path)
    text_parts = []
    for para in doc.paragraphs:
        if para.text.strip():
            text_parts.append(para.text)
    # Also extract tables
    for table in doc.tables:
        for row in table.rows:
            row_text = " | ".join(cell.text.strip() for cell in row.cells)
            if row_text.strip():
                text_parts.append(row_text)
    return "\n".join(text_parts)


def extract_text_from_txt(file_path: str) -> str:
    with open(file_path, "r", encoding="utf-8") as f:
        return f.read()


def extract_text(file_path: str) -> str:
    ext = Path(file_path).suffix.lower()
    if ext == ".pdf":
        return extract_text_from_pdf(file_path)
    elif ext in (".docx", ".doc"):
        return extract_text_from_docx(file_path)
    elif ext in (".txt", ".md", ".csv"):
        return extract_text_from_txt(file_path)
    else:
        raise ValueError(f"Unsupported file type: {ext}")

File: api/lib/services/file-processing.ts
/**
 * Extract text from uploaded documents (PDF, DOCX, TXT).
 */
import { extname } from "node:path";
import { readFile } from "node:fs/promises";

/**
 * Extract text from a PDF file using pdf-parse.
 */
export async function extractTextFromPdf(filePath: string): Promise<string> {
  const pdfParse = (await import("pdf-parse")).default;
  const buffer = await readFile(filePath);
  const data = await pdfParse(buffer);
  return data.text;
}

/**
 * Extract text from a DOCX file using mammoth.
 */
export async function extractTextFromDocx(filePath: string): Promise<string> {
  const mammoth = await import("mammoth");
  const buffer = await readFile(filePath);

  // mammoth.extractRawText gives us the text content without HTML markup
  const result = await mammoth.extractRawText({ buffer });
  return result.value;
}

/**
 * Extract text from a plain text file (txt, md, csv).
 */
export async function extractTextFromTxt(filePath: string): Promise<string> {
  const buffer = await readFile(filePath, "utf-8");
  return buffer;
}

/**
 * Dispatch to the correct extractor based on file extension.
 *
 * Supported formats: .pdf, .docx, .doc, .txt, .md, .csv
 */
export async function extractText(filePath: string): Promise<string> {
  const ext = extname(filePath).toLowerCase();

  switch (ext) {
    case ".pdf":
      return extractTextFromPdf(filePath);
    case ".docx":
    case ".doc":
      return extractTextFromDocx(filePath);
    case ".txt":
    case ".md":
    case ".csv":
      return extractTextFromTxt(filePath);
    default:
      throw new Error(`Unsupported file type: ${ext}`);
  }
}
